src/openrct2/core/Zip.cpp

```diff
@@ -106,66 +167,183 @@ class ZipArchive final : public IZipArchive
 
         auto source = zip_source_buffer(_zip, writeBuffer.data(), writeBuffer.size(), 0);
         auto index = GetIndexFromPath(path);
-        if (index == -1)
+        if (index)
         {
-            zip_add(_zip, path.data(), source);
+            zip_replace(_zip, *index, source);
         }
         else
         {
-            zip_replace(_zip, index, source);
+            zip_add(_zip, path.data(), source);
         }
     }
 
     void DeleteFile(std::string_view path) override
     {
         auto index = GetIndexFromPath(path);
-        zip_delete(_zip, index);
+        if (index)
+        {
+            zip_delete(_zip, *index);
+        }
+        else
+        {
+            throw std::runtime_error("File does not exist.");
+        }
     }
 
     void RenameFile(std::string_view path, std::string_view newPath) override
     {
         auto index = GetIndexFromPath(path);
-        zip_file_rename(_zip, index, newPath.data(), ZIP_FL_ENC_GUESS);
+        if (index)
+        {
+            zip_file_rename(_zip, *index, newPath.data(), ZIP_FL_ENC_GUESS);
+        }
+        else
+        {
+            throw std::runtime_error("File does not exist.");
+        }
     }
 
 private:
-    /**
-     * Normalises both the given path and the stored paths and finds the first match.
-     */
-    zip_int64_t GetIndexFromPath(std::string_view path) const
+    class ifilestream final : public std::istream
```

Comment:
Can't you adapt it from the existing FileStream?

Reply:
When and when should we not try to use the C++ classes where possible? Shouldn't we start replacing FileStream with fstream?

Reply:
Yes, what actually meant is we should stick to one thing.

Reply:
Okay, let this be the start of using built-in C++ streams.

Reply:
We should first check how well the performance would be, there are some optimizations in MemoryStream to better handle the desync debugging its still pretty slow so we have to see how that holds up at the end.